src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -2029,16 +2063,495 @@ def delete_user(self, user_name_or_id: Union[str, UUID]) -> None:
+                # regular user accounts here.
+                raise IllegalOperationError(
+                    f"The '{user.name}' account is a service account. "
+                    "Service accounts cannot be deleted as regular user "
+                    "accounts. Please use the service account interface "
+                    "instead."
+                )
             if user.name == self._default_user_name:
                 raise IllegalOperationError(
                     "The default user account cannot be deleted."
                 )
-
             self._trigger_event(StoreEvent.USER_DELETED, user_id=user.id)
 
             session.delete(user)
             session.commit()
 
+    # ----------------
+    # Service Accounts
+    # ----------------
+
+    def _get_service_account(
+        self, service_account_name_or_id: Union[str, UUID], session: Session
+    ) -> UserSchema:
+        """Helper method to fetch a service account by name or ID.
+
+        This is used on top of `_get_user_schema` to additionally filter
+        only user schemas that are service accounts. This is required because
+        in the DB, user accounts and service accounts are stored using the same
+        UserSchema for legacy reasons.
+
+        Args:
+            service_account_name_or_id: The name or ID of the service account
+                to fetch.
+            session: The database session to use for the query.
+
+        Returns:
+            The requested service account.
+
+        Raises:
+            KeyError: if the name or ID identifies a regular user account
+                instead of a service account.
+        """
+        account = self._get_user_schema(
+            service_account_name_or_id, session=session
+        )
+        if not account.is_service_account:
+            raise KeyError(
+                f"Unable to get service account with name or ID "
+                f"'{service_account_name_or_id}': No service account with "
+                "this ID found."
+            )
+
+        return account
+
+    def create_service_account(
+        self, service_account: ServiceAccountRequestModel
+    ) -> ServiceAccountResponseModel:
+        """Creates a new service account.
+
+        Args:
+            service_account: Service account to be created.
+
+        Returns:
+            The newly created service account.
+
+        Raises:
+            EntityExistsError: If a user or service account with the given name
+                already exists.
+        """
+        with Session(self.engine) as session:
+            # Check if a user or service account with the given name already
+            # exists
+            existing_user = session.exec(
+                select(UserSchema).where(
+                    UserSchema.name == service_account.name
+                )
+            ).first()
+            if existing_user is not None:
+                raise EntityExistsError(
```

Comment:
Is it necessary to enforce unique names for users + service accounts? While they share a single table in the database, it's not really something that users need to be aware of right?

We could still enforce unique user names and unique service account names but include the `is_service_account` flag in the unique constraint?

Reply:
I think you should not be able t create a service account with the same name as a user within a tenant simply for clarity, if a service account with name michael runs a pipeline, that would be confusing (even if we find soem way in the frontend to visually separate users from service accounts)

Reply:
We use the "user name" in various places (CLI, dashboard) as a visual hint for the owner of various resources. I agree with @AlexejPenner , it would be confusing to have overlapping names in this space, at least until the dashboard catches up with the service account concept. But you did make me aware of a case where this restriction is problematic: creating a service account with the name `stefan@zenml.io` would block me from having a user account with that same name. Especially in the ZenML Cloud context where usernames are externally managed and ZenML user accounts are created automatically based on SSO data, this would indeed be a problem.

I don't know how to solve this elegantly. Forcing the service account name to contain only alphanumeric characters would take care of our needs at least. Any ideas @schustmi @AlexejPenner ?

Reply:
I might just go ahead and do what @schustmi suggested and make `name`+`is_service_account` the unique constraint to avoid this new situation.

Reply:
This is the solution I went for, for future reference:
* user accounts and service accounts can have overlapping names
* calling `get_user` with an account name does not return a service account, even if one has the same name
* calling `get_service_account` never returns a user account
* however, calling `get_user` with a service account UUID will return that service account (for backwards-compatibility purposes).